User defined script parameters

meviktor opened the pull request:
Extending the current add-on with the feature of adding custom parameters (tags) to the menu items. Read the updated README.md for the high level description of the feature. I have added notes below which (hopefully) will help during the review.

### AddMenuItem.cs (dialog)
- Filling "the context selector combo box" by using the MenuItemContext enumeration.
	- Fixes a small bug: you were not able to select "Column" as the nem command / MenuItem's context
- ValidateInputs: the "context combo box" is not empty anymore when the dialog pops up, so no need to check if it has a value set. Checking if the name combo box has some value entered.
- Adding handlers for adding, editing, removing custom parameters.
- GetArgumentNamesInUse: you cannot add a parameter with the same name twice. Names of "context levels" (OBJECT, SERVER, VIEW, etc.) are also not permitted (to avoid any conflicts when making replacements in the T-SQL script before executing).

### OptionsDialogPage.cs
- Running validation for MenuItems when the user presses OK on the Options grid of SSMS

### SSMSObjectExplorerMenu.cs
- Asking values of user-defined parameters before making replacements in the T-SQL script

### MenuItem.cs
- Extending with a list of user-definable parameters
- Using MenuItemContext enumeration to determine the MenuItem's exact context in the Object Explorer
	- Fixes a small bug: now when adding/editing a MenuItem in the Options grid, the context can be selected from a dropdown list. This elminates a possible typo in the context name - it makes the command/MenuItem inaccessible from the right-click context menu in Object Explorer.
- Preserving compatibility with the XML exports of the previous versions by adding the property ContextDescription
- Adding validation for (runs through the user-defined parameters if any)

### AddUserDefinedParameter.cs (dialog)
- Dialog for adding/editing custom arguments
- Parameter name: anything, except: names of "context levels" (OBJECT, SERVER, VIEW, etc.) - every parameter of a MenuItem must have a unique name

### EnterUserDefinedArguments.cs (dialog)
- It's a container of ArgumentControls (one for each parametAddMenuItem.cs (dialog)
- Filling "the context selector combo box" by using the MenuItemContext enumeration.
	- Fixes a small bug: you were not able to select "Column" as the nem command / MenuItem's context
- ValidateInputs: the "context combo box" is not empty anymore when the dialog pops up, so no need to check if it has a value set. Checking if the name combo box has some value entered.
- Adding handlers for adding, editing, removing custom parameters.
- GetArgumentNamesInUse: you cannot add a parameter with the same name twice. Names of "context levels" (OBJECT, SERVER, VIEW, etc.) are also not permitted (to avoid any conflicts when making replacements in the T-SQL script before executing).

### OptionsDialogPage.cs
- Running validation for MenuItems when the user presses OK on the Options grid of SSMS

### SSMSObjectExplorerMenu.cs
- Asking values of user-defined parameters before making replacements in the T-SQL script

### MenuItem.cs
- Extending with a list of user-definable parameters
- Using MenuItemContext enumeration to determine the MenuItem's exact context in the Object Explorer
	- Fixes a small bug: now when adding/editing a MenuItem in the Options grid, the context can be selected from a dropdown list. This elminates a possible typo in the context name - it makes the command/MenuItem inaccessible from the right-click context menu in Object Explorer.
- Preserving compatibility with the XML exports of the previous versions by adding the property ContextDescription
- Adding validation for (runs through the user-defined parameters if any)

### AddUserDefinedParameter.cs (dialog)
- Dialog for adding/editing custom arguments
- Parameter name: anything, except: names of "context levels" (OBJECT, SERVER, VIEW, etc.) - every parameter of a MenuItem must have a unique name

### EnterUserDefinedArguments.cs (dialog)
- It's a container of ArgumentControls (one for each parameter of a MenuItem)
- This dialog is displayed to enter values for arguments, if any is defined for the MenuItem

### ArgumentControl.cs (control)
- Control for enetering parameter values
- The input field and the validation criteria adapts to the parameter's (data) type (checkbox for bit, numeric updown for int, etc.)

### EnumConverter.cs, MenuItemContextConverter.cs, UserDefinedArgumentTypeConverter.cs
- Used to provide conversion between the values of the MenuItemContext, UserDefinedParameterType enumerations and their string representations (provided for each value in a DescriptionAttribute)
	- This conversion is used to display these strings in the Options Grid, when the user selects context for a MenuItem (from a dropdown list)er of a MenuItem)
- This dialog is displayed to enter values for arguments, if any is defined for the MenuItem

### ArgumentControl.cs (control)
- Control for enetering parameter values
- The input field and the validation criteria adapts to the parameter's (data) type (checkbox for bit, numeric updown for int, etc.)

### EnumConverter.cs, MenuItemContextConverter.cs, UserDefinedArgumentTypeConverter.cs
- Used to provide conversion between the values of the MenuItemContext, UserDefinedParameterType enumerations and their string representations (provided for each value in a DescriptionAttribute)
	- This conversion is used to display these strings in the Options Grid, when the user selects context for a MenuItem (from a dropdown list)

meviktor committed:
Adding custom argument model and the add/edit dialog.

meviktor committed:
AddOrEditCustomArgument dialog: additional validation. Remove unnecessary using from CustomArgument.cs file.

meviktor committed:
Rename CustomArgument to UserDefinedArgument in types. Extending AddMenuItem window with controls for adding user-defined arguments.

meviktor committed:
Displaying user-defined arguments in Options grid for view and edit. Fixing data binding in AddOrEditUserDefinedArgument dialog.

meviktor committed:
Renaming AddOrEditUserDefinedArgument to AddUserDefinedArgument.

meviktor committed:
Removing editing scenaro from AddUserDefinedArgument dialog's code.

meviktor committed:
Refactoring & fixes:
- Renaming everything (types, variables) contains "argument", "arg" to "parameter", "param".
- Moving extensions under own folder/namespace.
- Fix for  AddMenuItem dialog: not saved user-defined parameters with the command.

meviktor committed:
Enabling user-defined parameters in T-SQL scripts. See details below.

AddMenuItem: passing parameters coming from (execution) context as part of parameter names in use.
AddUserDefinedParameter: refactored validation. Modified exception message in Parameter getter.
OptionsDialogPage: adding validation for MenuItems when user clicks OK on Options page. Changes will be discarded if user clicks on Cancel or closes the Options page.
SSMSObjectExplorerMenu: replacing user-defined parameters in T-SQL script of the MenuItem/command.
Utils: fix context parameter names.
Extensions: added ReplaceRange extension method. Removed unnecessary empty lines.
UIExtensions: added ToUserDefinedArgument extension method.
MenuItem: added validation.
UserDefinedArgument added.
ArgumentControl: new control for entering value for a user-defined parameter.
EnterUserDefinedArgument: new dialog for entering parameter values before executing the T-SQL script of the MenuItem/command.

meviktor committed:
UI fixes:
- AddMenuItem: making user-defined parameter listview larger & resizeable. Increasing window & control sizes.
- AddUserDefinedParameter: displaying window in center of parent, modify sytle.
- EnterUserDefinedArgument: displaying window in center of parent, modify sytle. Auto-sized flowLayoutPanel.

meviktor committed:
Making parameter name replacement case-insensitive.

meviktor committed:
Adding bit (boolean) parameter type.

meviktor committed:
Adding CustomList parameter type. Refactoring in validation.

meviktor committed:
Adding missing inital value of the CustomList collection - fixing not saving value list.

meviktor committed:
Arranging controls of ArgumentControl - providing more space between controls.

meviktor committed:
Turning MenuItem context into enumeration type. Fixing issues the below issues:
- "Server/Database/Table/Column" could not be selected on the AddMenuItem dialog as the new command's context.
- In the Options grid, Context property of the command/menu item could be edited via a text box. This was error prone - in case of a typo, the command disappeared from the Object Explorer context menu. Now it can be selected from a dropdown.

meviktor committed:
Providing description for UserDefinedParameterType values.

meviktor committed:
Changing listview's view mode in AddUserDefinedParameter dialog.

meviktor committed:
Fixing displayed text in labels, dialogs. Moving enum converters in individual files.

meviktor committed:
Enabling/disabling the dialog buttons instead of displaying error popup. Allowin editing parameters on the AddMenuItem dialog.

meviktor committed:
Fixing XML export / import:
- displaying the full path in the Context field of the exported XML
- accepting string containing the full path to parse the appropriate MenuItemContext value.

meviktor committed:
Refactoring, adding explanation comments.
Minor UI changes.
Fix the issue when validation threw an exception in case the user did not enter anything for parameter name (UserDefinedParameter.cs)

meviktor committed:
Small fix for displaying user-defined parameter's type in the add menu item dialog.

meviktor committed:
Updating README.

meviktor committed:
Merge branch 'main' into user-defined-script-parameters.

meviktor committed:
Fix AddMenuItem.png

brink-daniel commented on enums/UserDefinedArgumentType.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace SSMSObjectExplorerMenu.enums
+{
+    public enum UserDefinedParameterType : byte
```
Feature request. Could we please add DateTime as another option.

brink-daniel commented on EnterUserDefinedArguments.cs:
```diff
@@ -0,0 +1,56 @@
+namespace SSMSObjectExplorerMenu
+{
+    public partial class EnterUserDefinedArguments : Form
+    {
```
Bug. Parameter controls are too wide, forcing the horizontal scroll bar to always be displayed.

meviktor replied:
Hmm, interesting... I see no horizontal scrollbar. However, when I provide a long parameter name, the end of it is just not visible. To be honest, that's why I introduced the NAME_MAX_LENGTH constant in UserDefinedParameter.cs file. Maybe I should reduce the maximum length limit further.
Anyway, I will look into this and try to reproduce this horizontal scrollbar issue you mentioned.

meviktor commented on objects/UserDefinedParameter.cs:
```diff
@@ -0,0 +1,82 @@
+namespace SSMSObjectExplorerMenu.objects
+{
+    [TypeConverter(typeof(ExpandableObjectConverter))]
+    public class UserDefinedParameter
```
This crossed my mind also during development... would you do it mandatory to specify a default value each parameter, or it would be just an option?

brink-daniel commented on EnterUserDefinedArguments.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace SSMSObjectExplorerMenu
+{
+    public partial class EnterUserDefinedArguments : Form
```
I would simply make the form resizable and achor the flow panel to resize with form.

meviktor committed:
Fixing bug: user is not able to set negative value as an integer argument.

meviktor committed:
Removed the maximum size limit of the "Enter script arguments..." window. The flow panel and window now resize in sync.

meviktor committed:
Made argument controls as wide as their container (FlowLayout). Removed unused padding.

meviktor committed:
Datetime2, datetimeoffset user-defined arguments.

meviktor committed:
Default value for user-defined parameters: extending UserDefinedParameter class, added validation. Filling ArgumentControls with the default values if any.

brink-daniel commented on enums/MenuItemContext.cs:
```diff
@@ -0,0 +1,36 @@
+namespace SSMSObjectExplorerMenu.enums
+{
+    [Flags]
+    public enum MenuItemContext
```
When you create a new menu item, the context is not set correctly 
<img width="686" height="473" alt="image" src="https://github.com/user-attachments/assets/7f2350d6-2b60-4645-b12f-d2644ebf2b45" />

meviktor committed:
Using default parameters - setting NULL as default value missing + handling CustomList type.
- AddUserDefinedParameter.cs: using DefaultValueControl for setting default parameter value + refactoring.
- ArgumentControl.cs: refactoring + using plain textboxes instead of "placeholder textboxes" to enter datetime2, datetimeoffset values.
- Extensions.cs: not throwing exception for NULL values during validation.
- UIExtensions.cs: new extension method for fetching values from controls.
- UserDefinedParameter.cs: removed boolean field if the parameter has a default value (don't needed as every parameter must have a default value).
- Constants.cs: added format strings.
- DefaultValueControl.cs: custom control for entering default parameter values. Doesn't support setting NULL as default value yet.

meviktor committed:
Fix: can't store default value of bit type.

meviktor committed:
ObservableListView with window procedure overload. Remove unused control in EnterUserDefinedArguments dialog.

meviktor committed:
Managing observability of ListView via wrapper methods instead of window messages (window procedure). Fixes in DefaultValueControl regarding the default value dropdown for CustomList parameter type.

meviktor committed:
Small fixes:
- fix wrong casing after editing MenuItem parameters in AddMenuItem dialog
- validating values for MenuItem parameters: no restriction for nvarchar params (can be empty string or whitespace)
- datetime2, datetimeoffset default values: today midnight

meviktor committed:
Removing MenuItemContext enumeration, reverting related changes. Cleanup unused code.

meviktor committed:
Increase height of DefaultValueControl.

meviktor committed:
Updating README.

meviktor committed:
Merge 'upstream/main' into 'user-defined-script-parameters'.

meviktor committed:
Fix for alignment of UI elements on AddUserDefinedParameter dialog.

meviktor committed:
Moving custom controls under their own folder and namespace.
